core/src/main/java/org/apache/iceberg/catalog/TableCommit.java

```diff
@@ -0,0 +1,31 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.catalog;
+
+import org.apache.iceberg.TableMetadata;
+import org.immutables.value.Value;
+
+@Value.Immutable
+public interface TableCommit {
```

Comment:
Do we want this to be based on `TableMetadata` or do we want it based on changes?

I think we want it based on changes that come from `TableMetadata`. That way we don't need to diff `TableMetadata` objects. We can do that in `TableOperations` because that controls the metadata base. But here, we could be passed two objects that are completely unrelated. So it makes more sense for the caller to manage `base` and `updated` and to get the change set from updated.

Reply:
indeed this would allow passing two random `TableMetadata` objects, which isn't ideal. I've replaced `updated` with the actual metadata changes.
The `base` metadata here is only needed to build the actual `requirements` for the changes.

Reply:
Why include `TableMetadata` rather than both requirements and updates?

Reply:
that would be ideal, but currently `UpdateRequirement` is part of the REST package and defined inside `UpdateTableRequest`. We would have to extract that out from there, which I wanted to avoid in this PR